feat(rust): add timer to close secure channel

Alef-gabriel opened the pull request:
## Current behavior

Currently, when to establish a secure channel, two workers are created: decrypt and encrypt. However, when the initiator shuts down its connection without explicitly removing the workers, they remain active on the responder side. This can lead to issues such as performance degradation or memory consumption over time.

## Proposed changes
A timer to shut down the workers when a message is not received

Alef-gabriel committed:
feat(rust): add timer to close secure channel

Alef-gabriel commented on implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs:
```diff
@@ -93,6 +95,7 @@ impl Worker for HandshakeWorker {
             } else {
                 Err(IdentityError::UnknownChannelMsgDestination.into())
             };
+            self.flag.store(true, Ordering::Relaxed);
```
In this case why is this not working? if I put this line outside of this code block It's work normally.

davide-baldo replied:
I can speculate that the spawned task is checking the flag **before** the initialization completes, whereas if you move the flag.store in the upper block it would also be called during initilization.

davide-baldo replied:
probably starting from a `true` value would do the trick

Alef-gabriel commented:
@davide-baldo

davide-baldo commented on implementations/rust/ockam/ockam_identity/src/secure_channels/secure_channels.rs:
```diff
@@ -108,10 +112,27 @@ impl SecureChannels {
+        let addr = addresses.encryptor.clone();
+        tokio::spawn(async move {
+            let flag_value = Arc::clone(&flag);
+            while !flag_value.load(Ordering::Relaxed) {
```
You are already checking below:
```
if !flag_value.load(Ordering::Relaxed) {
    self_clone.stop_secure_channel(&ctx_clone, &addr).await.unwrap();
    break;
}
```
If the flag changes state just after you set it to false (highly unlikely by possible), the loop would exit in the main condition and `self_clone.stop_secure_channel(&ctx_clone, &addr).await.unwrap();` would not be called.

davide-baldo commented on implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs:
```diff
@@ -40,6 +41,7 @@ pub(crate) struct HandshakeWorker {
     role: Role,
     remote_route: Option<Route>,
     decryptor_handler: Option<DecryptorHandler>,
+    flag: Arc<AtomicBool>,
```
Every variable name is an opportunity you have to give hints to the reader about your intention, try to give a more meaningful name that reflects what you had in mind, even if imperfect.

davide-baldo commented:
I like it! The direction is excellent!

The only note I have is that using a boolean is somewhat fragile because you don't have any guarantee about when a task is actually run. For example, if some piece of the code misbehaves and blocks an asynchronous thread, every task that is using that thread will be suspended until the piece of code returns. In this case, the secure channel may be idle for the duration of the suspension (which could be minutes) and not detect it.

The most common solution would be to store the current timestamp (less performant) instead of storing the boolean. But being a simple timeout, even if we are not super precise, I think it's ok, the goal is to clean up resources, and it's doing that.

Alef-gabriel committed:
fix(rust): modify timeout logic

Alef-gabriel commented:
Hi @davide-baldo !
I'm not sure if it's suitable, but the security channel takes twice as long to close with timeout.

etorreborre committed:
style(rust): format code

etorreborre committed:
feat(rust): compile for no_std

etorreborre committed:
refactor(rust): rename is_idle_timeout to is_idle

since this is just a boolean, not the timeout itself
Also, moved the code resetting the boolean to a better place

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/tests/channel.rs:
```diff
@@ -1106,7 +1106,7 @@ async fn should_stop_encryptor__and__decryptor__in__secure_channel(
     assert_eq!(local_info.their_identity_id(), alice.identifier());
     assert_eq!("Hello, Bob!", msg.body());
-
+    dbg!(ctx.list_workers().await?);
```
Do you think it makes sense to remove `dbg!` after the test is passing?

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/tests/channel.rs:
```diff
@@ -1118,9 +1118,161 @@ async fn should_stop_encryptor__and__decryptor__in__secure_channel(
+    let bob = identities_creation.create_identity().await?;
+    let alice = identities_creation.create_identity().await?;
+
+    let bob_trust_policy = TrustIdentifierPolicy::new(alice.identifier());
```
This is not a needed detail for that test. Might just use `SecureChannelListenerOptions::new()`. But it's also OK as it is

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/tests/channel.rs:
```diff
@@ -1118,9 +1118,161 @@ async fn should_stop_encryptor__and__decryptor__in__secure_channel(
+
+    let msg = child_ctx.receive::<String>().await?;
+
+    let local_info = IdentitySecureChannelLocalInfo::find_info(msg.local_message())?;
```
I think these 2 lines are not needed here

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/tests/channel.rs:
```diff
@@ -1118,9 +1118,161 @@ async fn should_stop_encryptor__and__decryptor__in__secure_channel(
+
+    let mut additional_workers = ctx.list_workers().await?;
+    additional_workers.retain(|w| !initial_workers.contains(w));
+    let works_count = additional_workers.len();
```
Could we please, instead of relying on the number of workers, instead check if the specific encryptors and decryptors are on the list of workers? I think there are potentially a lot of workers on the system that can start/stop even though we don't do anything, which can make this test incorrect in the future.

SanjoDeundiak replied:
That is also potentially the reason why `should_stop_encryptor__and__decryptor__per__timeout` is failing on the CI

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/src/secure_channels/secure_channels.rs:
```diff
@@ -108,10 +112,32 @@ impl SecureChannels {
+        // start a background thread to observe the connection activity
+        // If a message is received before maximum_idle_time has passed then the connection is
+        // active. Otherwise, we close the secure channel
+        spawn(async move {
```
That will shut down the channel in the range of [maximum_idle_time, 2 * maximum_idle_time) from the last message. Not quite precise, but it should be fine. Could you please mention that detail in the doc-comment?
Also, mentioning that Secure Channels are shut down due to inactivity is worth mentioning somewhere in the doc-comments of the public API. Could you please add that?

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/src/secure_channels/secure_channels.rs:
```diff
@@ -108,10 +112,32 @@ impl SecureChannels {
+                    self_clone
+                        .stop_secure_channel(&ctx_clone, &addr)
+                        .await
+                        .unwrap();
```
Let's please handle the error gracefully, we generally want to avoid panics

SanjoDeundiak commented on implementations/rust/ockam/ockam_identity/src/secure_channel/listener.rs:
```diff
@@ -99,6 +98,7 @@ impl Worker for IdentityChannelListener {
             self.options.trust_context.clone(),
             None,
             None,
+            None,
```
What is the reason that we don't use idle flag here?

SanjoDeundiak reviewed (changes requested):
Nice approach. Left few comments. Also, could you also please clone the `idle` flag to the Encryptor worker? Because currently using the encryption side of the Secure Channel doesn't reset the flag, therefore if we only use one-side communtication, the Secure Channel will be shut down soon :)

SanjoDeundiak commented:
Hey @Alef-gabriel , do you plan to get back to this PR?
